language/move-binary-format/src/file_format.rs

```diff
@@ -651,20 +673,21 @@ impl AbilitySet {
     /// the same ability, the type arguments need `Store`
     pub fn polymorphic_abilities(
         declared_abilities: Self,
-        type_argument_abilities: impl IntoIterator<Item = Self>,
+        type_arguments: impl IntoIterator<Item = (Self, TypeParamKind)>,
     ) -> Self {
         // Conceptually this is performing the following operation:
         // For any ability 'a' in `declared_abilities`
         // 'a' is in the result only if
-        //   for all 'ti' in `type_argument_abilities`, a.required() is a subset of ti
+        //   for all 'ti' in `type_arguments` s.t. ti.kind is not phantom, a.required() is a subset of ti.constraints
         //
         // So to do this efficiently, we can determine the required_by set for each ti
         // and intersect them together along with the declared abilities
         // This only works because for any ability y, |y.requires()| == 1
-        type_argument_abilities
+        type_arguments
             .into_iter()
-            .map(|ty_arg_abilities| {
-                ty_arg_abilities
+            .filter(|(_, kind)| kind.is_non_phantom())
```

Comment:
ah, so it is used. I'd probably do `!usage.is_phantom()`. feels a bit more descriptive